flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ExecutionAttemptMappingProvider.java

```diff
@@ -38,32 +37,38 @@
     private final List<ExecutionVertex> tasks;
 
     /** The cached mapping, which would only be updated on miss. */
-    private final LinkedHashMap<ExecutionAttemptID, ExecutionVertex> cachedTasksById;
+    private Map<ExecutionAttemptID, ExecutionVertex> cachedTasksById;
 
     public ExecutionAttemptMappingProvider(Iterable<ExecutionVertex> tasksIterable) {
         this.tasks = new ArrayList<>();
         tasksIterable.forEach(this.tasks::add);
 
-        this.cachedTasksById =
-                new LinkedHashMap<ExecutionAttemptID, ExecutionVertex>(tasks.size()) {
-
-                    @Override
-                    protected boolean removeEldestEntry(
-                            Map.Entry<ExecutionAttemptID, ExecutionVertex> eldest) {
-                        return size() > tasks.size();
-                    }
-                };
+        this.cachedTasksById = new HashMap<>(tasks.size());
     }
 
     public Optional<ExecutionVertex> getVertex(ExecutionAttemptID id) {
-        if (!cachedTasksById.containsKey(id)) {
-            cachedTasksById.putAll(getCurrentAttemptMappings());
-            if (!cachedTasksById.containsKey(id)) {
-                // the task probably gone after a restart
-                cachedTasksById.put(id, null);
+        ExecutionVertex vertex = cachedTasksById.get(id);
+        if (vertex != null || cachedTasksById.containsKey(id)) {
+            return Optional.ofNullable(vertex);
+        }
+
+        return updateAndGet(id);
+    }
+
+    private Optional<ExecutionVertex> updateAndGet(ExecutionAttemptID id) {
+        synchronized (tasks) {
+            ExecutionVertex vertex = cachedTasksById.get(id);
+            if (vertex != null) {
```

Comment:
shouldn't we check `containsKey` here as well?

Reply:
Yes, sure. I missed it